docker/jupyter/custom/JupyterLockingExtension.js

```diff
@@ -0,0 +1,327 @@
+const dialog = require('base/js/dialog')
+const utils = require("base/js/utils")
+
+define(() => {
+    var modalOpen = false
+    var meta = {}
+
+    const syncIssueButtons = {
+        'Make a Copy': {
+            click: () => saveAs()
+        },
+        'Reload the workspace version and discard your changes': {
+            click: () => updateLocalCopyWithRemote()
+        }
+    }
+
+    const lockIssueButtons = {
+        'Run in Playground Mode': {
+            click: () => openPlaygroundMode(),
+            'class': 'btn-primary'
+        },
+        'Make a Copy': {
+            click: () => saveAs()
+        }
+    }
+
+    const noRemoteFileButtons = {
+        'Continue working': {
+            click: () => {},
+            'class': 'btn-primary'
+        }
+    }
+
+    const modeBannerId = "notification_mode"
+    const lockConflictTitle = "File is in use"
+    const syncIssueTitle = "File versions out of sync"
+    const syncIssueBody = "Your version of this file does not match the version in the workspace. What would you like to do?"
+    const syncIssueFatalBody = "This file was either deleted or never was stored with us."
+
+    //TODO URL resolution
+    const leoUrl = 'http://localhost:8080'
+        // const leoUrl = 'http://localhost:8081'
+    const welderUrl = leoUrl
+    const localizeUrl = welderUrl + '/localize'
+        // const checkMetaUrl = welderUrl + '/checkMeta'
+    const checkMetaUrl = welderUrl + '/objects/metadata'
+    const lockUrl = welderUrl + '/lock'
+    const lastLockedTimer = 60000 // in ms, should be 60000 in final PR
+
+    const headers = {
+        'Content-Type': 'application/json',
+        'Access-Control-Allow-Headers': '*'
+    }
+
+    const basePayload = {
+        // mode: "no-cors",
+        headers: headers
+    }
+
+    //TODO resolve these links
+    const jupyterBaseUrl = "http://localhost:8000/"
+    const jupyterContentsAPIUrl = jupyterBaseUrl + "api/contents/"
+
+    const getLocalPath = () => {
+        return {
+            localObjectPath: Jupyter.notebook.notebook_path
+        }
+    }
+
+    function init() {
+        checkMeta()
+        initSyncMaintainer()
+    }
+
+    async function initSyncMaintainer() {
+        var syncMaintainer = setInterval(() => {
+            checkMeta()
+        }, lastLockedTimer)
+
+        window.onbeforeunload(function() {
+            clearInterval(syncMaintainer)
+        })
+    }
+
+    async function checkMeta() {
+        const payload = Object.assign(basePayload, {
```

Comment:
This mutates basePayload - I don't think that's what you want. splay shallow copy is probably good here:


```
fetch(url, {
  ...basePayload,
  body: JSON.stringify(..
}
```